constantine/math_compiler/experimental/backends/cuda.nim

```diff
@@ -114,11 +130,133 @@ proc genFunctionType*(typ: GpuType, fn: string, fnArgs: string): string =
+  ## called in the scope is added to `fnTab`. This is a form of dead code elimination.
+  case n.kind
+  of gpuCall:
+    let fn = n.cName
+    if fn in ctx.allFnTab:
+      # Check if any of the parameters are pointers (otherwise non generic)
+      if fn notin ctx.fnTab: # function not known, add to `fnTab` (i.e. avoid code elimination)
+        let fnCalled = ctx.allFnTab[fn]
+        ctx.fnTab[fn] = fnCalled
+        # still "scan for functions", i.e. fill `fnTab` from inner calls
+        for ch in fnCalled:
+          ctx.scanFunctions(ch)
+      # else we don't do anything for this function
+    # Harvest functions from arguments to this call!
+    for arg in n.cArgs:
+      ctx.scanFunctions(arg)
+  else:
+    for ch in n:
+      ctx.scanFunctions(ch)
+
+proc getFieldType(t: GpuType, field: GpuAst): GpuType =
+  ## Returns the type of the field. `t` must be an object or generic instantiation.
+  ## `field` must be an ident.
+  doAssert field.kind == gpuIdent, "Field is not an ident: " & $field
+  doAssert t.kind in [gtObject, gtGenericInst]
+  let flds = if t.kind == gtObject: t.oFields
+               else: t.gFields
+  result = GpuType(kind: gtInvalid)
+  for f in flds:
+    if f.name == field.ident():
+      return f.typ
+
+proc getType(ctx: var GpuContext, arg: GpuAst, typeOfIndex = true): GpuType =
+  ## Tries to determine the underlying type of the AST.
+  ##
+  ## If `typeOfIndex` is `true`, we return the type of the index we access. Otherwise
+  ## we return the type of the array / pointer.
+  ##
+  ## NOTE: Do *not* rely on this for `mutable` or `implicit` fields of pointer types!
+  template dfl(): untyped = GpuType(kind: gtInvalid)
+  case arg.kind
+  of gpuIdent: arg.iTyp
+  of gpuAddr: GpuType(kind: gtPtr, to: ctx.getType(arg.aOf))
+  of gpuDeref:
+    let argTyp = ctx.getType(arg.dOf)
+    doAssert argTyp.kind == gtPtr
+    argTyp.to
+  of gpuCall: dfl()
+  of gpuIndex:
+    let arrType = ctx.getType(arg.iArr)
+    if typeOfIndex:
+      case arrType.kind
+      of gtPtr:   arrType.to
+      of gtUA:    arrType.uaTo
+      of gtArray: arrType.aTyp
+      else: raiseAssert "`gpuIndex` cannot be of a non pointer / array type: " & $arrType
+    else:
+      arrType
+  of gpuDot:
+    let parentTyp = ctx.getType(arg.dParent)
+    parentTyp.getFieldType(arg.dField)
+  of gpuLit: arg.lType
+  of gpuBinOp: dfl() ## XXX: store resulting type of `gpuBinOp`!
+  #of gpuBlock: arg.statements[^1].getType()
+  of gpuPrefix: ctx.getType(arg.pVal)
+  of gpuConv: arg.convTo
+  of gpuCast: arg.cTo # ident of the thing we cast
+  else:
+    raiseAssert "Not implemented to determine type from node: " & $arg
+
+proc makeCodeValid(ctx: var GpuContext, n: var GpuAst) =
+  ## Addresses other AST patterns that need to be rewritten on CUDA. Aspects
+  ## that are rewritten include:
+  ##
+  ## - `Index` of `Deref` of `Ident` needs to be rewritten to `Index` of `Ident` if the
```

Comment:
Concrete examples would help maintenance, debugging or generalizing in the future

Reply:
I have test cases for (I think) all of these (incl the WebGPU ones) as part of the Lita repository. I can move those over here, I suppose.